### DISCUSSION TITLE
ZFS how to mount an existing pool

### DISCUSSION THREAD

**Discussion by runningman84** [siderolabs/talos#8488] [General]
Has anyboy tried to use an existing zfs volume using talos? 

I get errors like this 10.0.30.83: cannot mount '/tank': failed to create mountpoint: Read-only file system

zfs tends to mount volumes at top level which is readonly in talos

I also tried to mount it using the machinedisk config https://www.talos.dev/v1.6/reference/configuration/#machinedisk
 which resulted in talos failing to boot completely.

What is the right way to deal with existing zfs disks? In my case these are usb attached drives where it also does not make sense to access them using a device name like /dev/sdx. It is also possible to mount them via some uuid or something like that?

**Comment by smira** [comment]
Talos rootfs is read-only, you can't create a directory there. You can mount under `/var` though, e.g. `/var/mnt/tank` or something similar.

**Comment by runningman84** [comment]
okay I used another ubuntu machine to change things like this
````
zfs set mountpoint=/var/mnt/tank tank
````
Now it works:
````
➜  talos talosctl ls /var/mnt/tank
NODE         NAME
10.0.30.83   .
10.0.30.83   alice
10.0.30.83   bob
````

I think for production use talos would need some kind of zfs api to interact with all the zfs magic.

The normal mount stuff does not apply here....

**Comment by runningman84** [comment]
or would you advice to use some root pod with zfs utils to do these changes?

**Comment by frezbo** [comment]
> or would you advice to use some root pod with zfs utils to do these changes?

yes, use a privileged container for these operations

**Comment by saolof** [comment]
You can use the -m legacy option when creating the pool to prevent it from trying to mount at a default location, and rely on a CSI driver like the openebs one to actually mount volumes

**Comment by henyxia** [comment]
A little addition for previous setups with NFS shares.
It seems that just setting `zfs set mountpoint=/var/mnt/tank tank` do not do the trick directly as the update is blocked by NFS server not being available:
```
/ # chroot /host zfs set mountpoint=/var/mnt/tank tank
cannot share 'tank: system error': NFS share creation failed
cannot share 'tank/xxx: system error': NFS share creation failed
cannot share 'tank/yyy: system error': NFS share creation failed
```
And `ls` showing an empty folder.
To fix this, before running the mountpoint update, just disable the NFS options with:
```
/ # chroot /host zfs set sharenfs='off' tank
```
All other shares should also be disabled, which can be checked using:
```
chroot /host zfs get sharenfs
```
Then, the mountpoint update will work like a charm.
